Automattic/woocommerce-payments
# Terminal In-Person Payments auto-renewing subscription support

Fixes #

#### Changes proposed in this Pull Request

- Add tokens to order and user when collecting a terminal payment for an order containing subscription products.
- Undeprecate `payments/orders/{order_id}/create_customer` endpoint for use in taking payments for subscriptions.


## Experimental Terminal subscription support


When we take payment for an order with subscriptions using the IPP terminal, the subscription renewals are not set up.

This PR address that need, by saving the tokens and preparing the subscription for auto-collection, where we are capable of doing that (country limitations apply.)

[This approach uses PaymentIntents which are set up with `setupFutureUsage = "off_session"` ](https://docs.stripe.com/terminal/features/saving-cards/save-after-payment?terminal-sdk-platform=ios#client-side).

The clients (e.g. mobile apps) are responsible for setting up the subscription, setting up the Customer for the order, and then calling the (now-undeprecated) `create_customer` endpoint to fetch the customer ID. They will attach this to the payment method, along with the setup future usage parameter, and then collect the payment method.

This is submitted as usual to `capture_terminal_payment`.

The changes to the `capture_terminal_payment` endpoint check for any subscriptions related to the order, and for a `generated_card` payment method to use for renewals on the captured payment intent. If both of these are present, it sets up the renewals by adding the reusable payment method to the user, and to each subscription relating to the order, and finally disables manual renewal (where not compulsory.)



The code relies on WCS being present, and checks for each function it uses before doing any work, so we don't "half-activate" a subscription if functions are not available.

If the `generated_card` is not present, the endpoint succeeds anyway, even if there are subscriptions on the order. This is because the initial payment will still have been taken, and Subscriptions will handle required manual renewals without risk of the merchant losing money. We could consider adding to the success response in this case so that the apps can alert the merchant that renewals won't be set up until another payment method is provided.

A thought – We could consider adding a pre-capture check for orders with subscription products, but no subscriptions created, and error if that occurs... so we don't capture payments when something has gone wrong with the previous step. I'm not sure whether that would be dealing with the concerns of subscriptions too much, but I suppose the check could be calling another function in WCS to see whether the order is payable yet...?



#### Testing instructions




Set up the iOS app (in a simulator if needed) https://github.com/woocommerce/woocommerce-ios/pull/12382 – this allows payment for subscriptions, sets `off_session` future usage, and has somewhere to put the customer ID.

* Add an order in WP admin containing one or more subscription products, with a user selected for the customer.
* Make a POST request to `orders/{order_id}/subscriptions` to create the subscription(s)
* Make a POST request to `payments/orders/112/create_customer` to create the customer in Stripe
* Modify the app code to add the customer ID from the previous step. In [`PaymentIntentParameters.build(_:with:)`](https://github.com/woocommerce/woocommerce-ios/blob/62e1a86654af8d8576f7d2f4e9f28ec059adf925/Hardware/Hardware/CardReader/StripeCardReader/PaymentIntentParameters%2BStripe.swift#L39), you'll see there's a `cus_xyz` ID hardcoded – replace this with the ID from the previous step.
* Set the `-simulate-stripe-card-reader` option in the app build scheme
* Build and run
* Go to the Orders tab in the app
* Select the order
* Tap `Collect Payment`
* Tap `Card`
* Wait for the card payment to be collected
* Check back in the Subscriptions page on WC admin – observe that the subscription has been activated, and in the database see that the payment method has been set.

https://github.com/Automattic/woocommerce-payments/assets/2472348/a1a933bc-e6ff-4370-9c84-b192d06cb388


-------------------


**Post merge**
